File: 序号/743. 网络延迟时间.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        g=[[] for _ in range(n+1)]
        for u,v,w in times:
            g[u].append((v,w))
        q=[(0,k)]
        dis=[math.inf]*(n+1)
        dis[k]=0
        vis=set()
        while q:
            _,u = heapq.heappop(q)
            if u in vis:continue
            vis.add(u)
            for v,w in g[u]:
                if dis[v]>dis[u]+w:
                    dis[v]=dis[u]+w
                    heapq.heappush(q,(dis[v],v))
                    
        maxD = 0
        for i in range(1,n+1):
            if dis[i]==math.inf:
                return -1
            maxD = max(maxD,dis[i])
        return maxD
        

File: 序号/743. 网络延迟时间.kt
class Solution {
  fun networkDelayTime(times: Array<IntArray>, n: Int, k: Int): Int {
    val g = Array(n+1){ mutableListOf<IntArray>() }
    for ((u,v,w) in times){
      g[u].add(intArrayOf(v,w))
    }
    val dis = IntArray(n+1){Int.MAX_VALUE}
    dis[k]=0
    val q = PriorityQueue<IntArray>(){a,b-> a[0]-b[0]}
    val vis = BooleanArray(n+1)
    q.add(intArrayOf(0,k))
    while(q.isNotEmpty()){
      val (_,u) = q.poll()!!
      vis[u]=true
      for((v,w) in g[u]){
        if (dis[v]>dis[u]+w){
          dis[v]=dis[u]+w
          q.add(intArrayOf(dis[v],v))
        }
      }
    }
    var ans = 0
    for (i in 1 until dis.size){
      if (dis[i]==Int.MAX_VALUE) return -1
      if (dis[i]>ans) ans = dis[i]
    }
    return ans
  }
}